No more leaking of mounttype via metadata or icon in public shares.

rullzer opened the pull request:
Patch for #13789

There are no unit tests. I could not find unit tests for the ajax/list.php stuf, which makes sense since it is far from trivial to test.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c3a4458f-cc63-49e7-ad30-1f3e7267bf52) was created.

LukasReschke commented:
I'd prefer if we could at this point also move to an AppFramework controller here, because then we're able to test this ;-)

See https://github.com/owncloud/core/tree/master/apps/files_sharing/lib/controllers and https://github.com/owncloud/core/blob/master/apps/files_sharing/appinfo/routes.php. Some easier examples how to use the AppFramework can be found at http://doc.owncloud.org/server/8.0/developer_manual/app/tutorial.html

Interested to adjust this as well? – If you need some hints or help just ask and you shall receive :smile:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c3a4458f-cc63-49e7-ad30-1f3e7267bf52) completed: **No issues found**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c3a4458f-cc63-49e7-ad30-1f3e7267bf52) completed: **No issues found**

rullzer commented:
Sure, I will have a look

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4f8a3522-4411-4931-9f47-ac3aeaf69151) was created.

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
I'd remove that. It's always better to have the CSRF check executed even if not technically 100% needed as a defense-in-depth against yet unknown vectors.

If you use the jQuery AJAX functions ownCloud adds the adder automatically.

rullzer commented:
I just added some commits to move to the appframework for the ajax/list.php
In order to do this the helper class needed to be non static.

I started with writing unit tests. Maybe more a required? So we can verify that everything keeps working!

I removed the mixing of GET and POST. Since as far as I could tell we were using the session to authenticate with passwords anyway. As least in my debug sessions I never came across a password begin posted.

Feedback is welcome!

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+	 * @param string sort
+	 * @param string sortdirection
+	 */
+	public function getList($t, $dir, $sort = 'name', $sortdirection = 'desc') {
```
Please add on multiple lines and replace `$t` with something like `$token` to make it easier to understand :)

rullzer replied:
I agree that woul be better. However all the getList stuff uses t=.... chaning this to token breaks routing... or has to be done in a lot of places at once (a lot of which are curerntly not unit tested)

LukasReschke replied:
*sigh*

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
```
New header please, see https://github.com/owncloud/core/pull/14426

LukasReschke commented on apps/files_sharing/tests/controller/ajaxcontroller.php:
```diff
@@ -0,0 +1,206 @@
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
```
New header please, see https://github.com/owncloud/core/pull/14426

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+		}
+		$data['permissions'] = $permissions;
+
+		return new JSONResponse(
```
Let's use a `DataResponse` instead which is more flexible as we could add more response formatters in the future.

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+		$absdir = Filesystem::normalizePath($absdir);
+
+		if (!Filesystem::is_dir($absdir . '/')) {
+			return new JSONRespone(
```
Also DataResponse

LukasReschke replied:
It's also a typo in here. Which IDE do you use? PHPStorm should show such errors directly to you :)

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+		} catch (AuthenticationException $e) {
+			return new JSONResponse(array(), Http::STATUS_FORBIDDEN);
+		} catch (\Exception $e) {
+			return new JSONResponse(array(), Http::STATUS_INTERNAL_SERVER_ERROR);
```
DataResponse as stated in the other comments

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+	 * @param string t
+	 * @param string dir
+	 * @param string sort
+	 * @param string sortdirection
```
PHPDoc requires to use `$` in front of variable names

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+use OCP\IRequest;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\JSONResponse;
+use OCA\Files_Sharing\Helper2;
```
Not existing file? – Missed to commit? – Might also make sense to rename this into some better name :)

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+			unset($entry['directory']);
+			// do not disclose share owner
+			unset($entry['shareOwner']);
+			// do not disclose if somethign is a remote shares
```
`something`

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+		}
+
+		$linkItem = $data['linkItem'];
+		$absdir = $data['realPath'];
```
While at it rename to `$absDir` for better readability.

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+
+	/** @var AppConfig */
+	protected $appConfig;
+	/** @var OCP\Session */
```
That is an `ISession`

LukasReschke commented on apps/files_sharing/appinfo/routes.php:
```diff
@@ -30,7 +30,10 @@
 		'ExternalShares' => ['url' => '/api/externalShares'],
-	]
+	],
+	'routes' => array(
```
I'd prefer to stay consistent and use `[` (array syntax for PHP >= 5.4)

LukasReschke replied:
Same for other usages of `array` in the controller below.

LukasReschke commented on apps/files_sharing/appinfo/app.php:
```diff
@@ -41,6 +41,8 @@
 
 // Exceptions
 OC::$CLASSPATH['OCA\Files_Sharing\Exceptions\BrokenPath'] = 'files_sharing/lib/exceptions.php';
+OC::$CLASSPATH['OCA\Files_Sharing\Exceptions\InvalidTokenException'] = 'files_sharing/lib/exceptions.php';
```
:see_no_evil:

rullzer replied:
It seems this line is only required to keep phpunit happy... So I probably need to add it somewhere else s phpunit is happy.

LukasReschke commented on apps/files_sharing/application.php:
```diff
@@ -67,6 +68,14 @@ public function __construct(array $urlParams=array()){
+			return new AjaxController(
+				$c->query('AppName'),
+				$c->query('Request'),
+				$server->getAppConfig(),
```
@DeepDiver1975 @MorrisJobke  What is the state here? – Should people use IConfig or IAppConfig or whatever?

LukasReschke commented:
Thanks @rullzer. Great start! :smile:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4f8a3522-4411-4931-9f47-ac3aeaf69151) completed: **6 new issues, 18 updated code elements**

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+
+namespace OCA\Files_Sharing\Controllers;
+
+use OC;
```
Really required?

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,133 @@
+use OCP\AppFramework\Http\JSONResponse;
+use OCA\Files_Sharing\Helper2;
+use OC\Files\Filesystem;
+use OC\AppConfig;
```
Private API :see_no_evil:  - please try to use only functions in `OCP` if possible :)

rullzer replied:
This is a leftover from the original list.php
I did a quick search but could not find a public API for this. Did I miss anything or is it just not available?

LukasReschke replied:
Use IConfig->getAppValue etc…

rullzer replied:
Ah yes that one I know.
I ment for OC\Files\Filesystem::normalizePath

LukasReschke replied:
Ahh… Yes… Let me look if this is actually required here still or if those checks are duplicated… Later ;-)

rullzer replied:
I noticed that in helper2 there are more uses of the OCP classes... but I think it is unavoidable there

LukasReschke commented on apps/files_sharing/application.php:
```diff
@@ -67,6 +68,14 @@ public function __construct(array $urlParams=array()){
 				$c->query('ExternalManager')
 			);
 		});
+		$container->registerService('AjaxController', function(SimpleContainer $c) use ($server) {
```
I'd go with another not so generic name to make more clear what this controller is for. Maybe `FileListController` or something like that…?

rullzer replied:
I'm fine with that. However I named it Ajax controller to eventually move all the stuff from the ajax dir to a testable controller. And then a more generic name seems better.

I'm fine either way.

LukasReschke commented on apps/files_sharing/tests/controller/ajaxcontroller.php:
```diff
@@ -0,0 +1,206 @@
+	/** @var string */
+	private $oldUser;
+	/** @var string */
+	private $oldSERVER;
```
PHP is case-sensitive when it comes to variables, below you used `oldServer` and here it is `oldSERVER`

(though the code might have worked before as well because PHP would magically have created it :see_no_evil:)

LukasReschke commented on apps/files_sharing/tests/controller/ajaxcontroller.php:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	public function testInvalidAuth() {
+		$linkItem = \OCP\Share::getShareByToken($this->token, false);
```
`$linkItem` is unused?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7df5945d-f6a2-4438-8ac1-d58209be936c) was created.

rullzer commented:
@LukasReschke thanks for the feedback. I already made a bunch of commits to fix some of them. The others I'll have a look at later.

Commits will be nicely merged eventually of course :)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7df5945d-f6a2-4438-8ac1-d58209be936c) completed: **18 new issues, 22 updated code elements**

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+
```
Author and Co. missing

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+
+		$linkItem = \OCP\Share::getShareByToken($token, is_null($password) ? false : true);
+		if($linkItem === false || ($linkItem['item_type'] !== 'file' && $linkItem['item_type'] !== 'folder')) {
+			\OC_Log::write('core-preview', 'Passed token parameter is not valid', \OC_Log::DEBUG);
```
Can use ILogger which could get injected

rullzer replied:
Ah yes that was still on my todo list. Hopefully later tonight

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+		$path = null;
+		if (isset($rootLinkItem['uid_owner'])) {
+			\OCP\JSON::checkUserExists($rootLinkItem['uid_owner']);
+			\OC_Util::tearDownFS();
```
JSON functions are deprecated

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+		}
+
+		if ($path === null) {
+			\OCP\Util::writeLog('share', 'could not resolve linkItem', \OCP\Util::DEBUG);
```
ILogger

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+		}
+
+		if (isset($linkItem['share_with']) && (int)$linkItem['share_type'] === \OCP\Share::SHARE_TYPE_LINK) {
+			if (!self::authenticate($linkItem, $password)) {
```
$this?

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+	 * @param string $relativePath optional path relative to the share
+	 * @param string $password optional password
+	 *
+	 * @throws InvalidTokenException
```
Can throw other exceptions as well

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,148 @@
+					return false;
+				}
+			} else {
+				\OCP\Util::writeLog('share', 'Unknown share type '.$linkItem['share_type']
```
ILogger

rullzer commented:
@LukasReschke thanks for the comments. i have to admit helper2.php is still mostly a copy of helper.php.. More updates will follow (hopefully) soon.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/8d594869-e083-4c9d-bf36-3742e815f5aa) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cc1dda66-7872-4d41-b706-45c7d7c9a6ee) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16f6361a-689e-4fb9-b5f1-166b65446714) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16f6361a-689e-4fb9-b5f1-166b65446714) completed: **15 new issues, 23 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dc6b354f-1d83-4135-9cb1-1c18839f164c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dc6b354f-1d83-4135-9cb1-1c18839f164c) completed: **11 new issues, 24 updated code elements**

LukasReschke commented:
Rebase :)

LukasReschke commented on apps/files_sharing/appinfo/routes.php:
```diff
@@ -30,6 +31,9 @@
 		'ExternalShares' => ['url' => '/api/externalShares'],
+	],
+	'routes' => [
+		['name' => 'ajax#get_list', 'url' => '/ajax/list.php', 'verb' => 'GET'],
```
Maybe time to get rid of that redundant `.php` as well and change it in all JS files? :see_no_evil:

rullzer replied:
Well, that would be nicer. But would require the other ajax calls to be rewritten as well. So once that is done it seems like a good idea.

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,147 @@
+	 * @param string $dir Subdirectory requested
+	 * @param string $sort Sort attribute
+	 * @param string $sortdirection Sort direction
+	 */
```
`@return DataResponse`

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,147 @@
+		$data['dirToken'] = $linkItem['token'];
+
+		$permissions = $linkItem['permissions'];
+		if ($this->appConfig->getValue('core', 'shareapi_allow_public_upload', 'yes') === 'no') {
```
Always use IConfig if possible please :)

LukasReschke commented on apps/files_sharing/lib/controllers/ajaxcontroller.php:
```diff
@@ -0,0 +1,147 @@
+			$formattedFiles[] = $entry;
+		}
+
+		//$data['directory'] = $relativePath; <-- where does this come from?
```
https://github.com/owncloud/core/pull/14341/files#diff-0c38edccfcc1f9aa0e2d2c9a16734085L41

LukasReschke commented on apps/files_sharing/lib/controllers/sharecontroller.php:
```diff
@@ -1,7 +1,7 @@
 /**
  * @author Bjoern Schiessle <schiessle@owncloud.com>
- * @author Georg Ehrke <georg@owncloud.com>
+ * @author Georg Ehrke <georg@ownCloud.com>
```
Unrelated?

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,128 @@
+	/** @var ILogger */
+	protected $logger;
+
+	public function __construct(ISession $session,
```
PHPDoc please.

LukasReschke commented on apps/files_sharing/lib/helper2.php:
```diff
@@ -0,0 +1,128 @@
+	 * @throws LinkItemException
+	 * @throws ItemTypeException
+	 * @throws AuthenticationException
+	 */
```
![screen shot 2015-02-28 at 21 40 40](https://cloud.githubusercontent.com/assets/878997/6428004/7673acfe-bf92-11e4-9baf-6098c73a48ce.png)

LukasReschke commented on apps/files_sharing/tests/controller/ajaxcontroller.php:
```diff
@@ -0,0 +1,199 @@
+/**
+ * @package OCA\Files_Sharing\Controllers
+ */
+class AjaxControllerTest extends \PHPUnit_Framework_TestCase {
```
New tests should extend `\Test\TestCase`

LukasReschke commented on apps/files_sharing/tests/controller/ajaxcontroller.php:
```diff
@@ -0,0 +1,199 @@
+use OCP\Share;
+
+/**
+ * @package OCA\Files_Sharing\Controllers
```
Namespace looks wrong?

LukasReschke commented:
Looks already really good otherwise! – Thanks!!! :rocket: :crown:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d93c6b61-6a46-4c99-b22d-eaf56057853f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d93c6b61-6a46-4c99-b22d-eaf56057853f) completed: **13 new issues, 24 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d794a294-d590-424c-adbf-3dd18cfaf726) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d794a294-d590-424c-adbf-3dd18cfaf726) completed: **21 new issues, 29 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/42e01ba7-85f3-4beb-b306-75603c520bf0) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16b0c836-c885-475e-8efc-bb2b1ea65d24) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/42e01ba7-85f3-4beb-b306-75603c520bf0) completed: **21 new issues, 29 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16b0c836-c885-475e-8efc-bb2b1ea65d24) completed: **20 new issues, 30 updated code elements**

rullzer commented:
@PVince81 @MorrisJobke can I ask you both to take a look?

MorrisJobke commented:
I just monkey tested this. I shared some files and folders (internal and public). Tried to access them. Looked at the logs. I couldn't find anything suspicious.

I'm fine with that, but I can't judge on the code.

:+1:

PVince81 commented:
Needs rebase

MorrisJobke commented:
rebase needed :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/939ce790-111c-419d-b880-665040b0358c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/939ce790-111c-419d-b880-665040b0358c) completed: **18 new issues, 30 updated code elements**

MorrisJobke commented:
@owncloud-bot retest this please

rullzer commented:
Jenkins is happy. See #15279 
Please review.

MorrisJobke commented:
@LukasReschke @icewind1991 Review please :)

LukasReschke commented:
I really appreciate and like the change but as we have freeze tomorrow and this change is very huge and touches magic sharing code I'd appreciate to re-evaluate this again this for 8.2.

Since the touched bits shouldn't change much (if at all) until then I think this should not result in a rebase horror trip :)

rullzer commented:
Fine with me. It should be carefully reviewed since the rebase did create some conflicts. 

But we can merge it once we start working on 8.2.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/50c660df-4823-4eb7-b319-6131d9342dc7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/50c660df-4823-4eb7-b319-6131d9342dc7) completed: **12 new issues, 30 updated code elements**

MorrisJobke commented:
@rullzer rebase needed

PVince81 commented:
@rullzer great work, but we'd like to get rid of ajax/*.php and have the web UI use WebDAV instead. See #12353 

Is there anything from this PR that we could still keep regardless ?

PVince81 commented:
Reminds me I need to find a way to return the mount type through PROPFIND...

rullzer committed:
Do not leak mounttype of a public share

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ee1ea347-acd4-41c9-9a9b-42bac186432e) was created.

rullzer commented:
@PVince81 not really then. Or at least not something that requires major rebasing... I branched that original PR and will work on that when the todo list is smaller.

This PR is now tiny. But it does make sure the info is not leaked. I'm fine with not merging this as long as we keep in mind to do this properly for the webdav stuff :)

PVince81 commented:
Nice, I like its simplicity.

:+1: let's merge this for now

MorrisJobke commented:
Tested and works :*1:

MorrisJobke commented:
:+1:

PVince81 commented:
Here is the same for the public webdav endpoint: https://github.com/owncloud/core/pull/17605
